src/core/thread/router_table.hpp

```diff
@@ -0,0 +1,338 @@
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef ROUTER_TABLE_HPP_
+#define ROUTER_TABLE_HPP_
+
+#include "common/encoding.hpp"
+#include "common/locator.hpp"
+#include "mac/mac_frame.hpp"
+#include "thread/mle_constants.hpp"
+#include "thread/thread_tlvs.hpp"
+#include "thread/topology.hpp"
+
+namespace ot {
+
+#if OPENTHREAD_FTD
+
+class RouterTable : public InstanceLocator
+{
+public:
+    /**
+     * This class represents an iterator for iterating through entries in the router table.
+     *
+     */
+    class Iterator : public InstanceLocator
+    {
+    public:
+        /**
+         * This constructor initializes an `Iterator` instance to start from beginning of the router table.
+         *
+         * @param[in] aInstance  A reference to the OpenThread instance.
+         *
+         */
+        Iterator(Instance &aInstance);
+
+        /**
+         * This method resets the iterator to start over.
+         *
+         */
+        void Reset(void);
+
+        /**
+         * This method indicates if the iterator has reached the end of the list.
+         *
+         * @retval TRUE   The iterator has reached the end of the list.
+         * @retval FALSE  The iterator currently points to a valid entry.
+         *
+         */
+        bool IsDone(void) const { return (mRouter == NULL); }
+
+        /**
+         * This method advances the iterator.
+         *
+         * The iterator is moved to point to the next entry.  If there are no more entries matching the iterator
+         * becomes empty (i.e., `GetRouter()` returns `NULL` and `IsDone()` returns `true`.
```

Comment:
nit: missing `)` at the end. (checked and it's missing from `ChildTable` as well. :smiley: )